File: comscoreanalytics/src/main/java/com/bitmovin/player/integrations/comscoreanalytics/ComScoreBitmovinAdapter.java
package com.bitmovin.player.integrations.comscoreanalytics;

import android.util.Log;

import com.bitmovin.player.BitmovinPlayer;
import com.bitmovin.player.api.event.data.AdFinishedEvent;
import com.bitmovin.player.api.event.data.AdStartedEvent;
import com.bitmovin.player.api.event.data.ErrorEvent;
import com.bitmovin.player.api.event.data.PausedEvent;
import com.bitmovin.player.api.event.data.PlaybackFinishedEvent;
import com.bitmovin.player.api.event.data.PlayingEvent;
import com.bitmovin.player.api.event.data.SourceUnloadedEvent;
import com.bitmovin.player.api.event.listener.OnAdFinishedListener;
import com.bitmovin.player.api.event.listener.OnAdStartedListener;
import com.bitmovin.player.api.event.listener.OnErrorListener;
import com.bitmovin.player.api.event.listener.OnPausedListener;
import com.bitmovin.player.api.event.listener.OnPlaybackFinishedListener;
import com.bitmovin.player.api.event.listener.OnPlayingListener;
import com.bitmovin.player.api.event.listener.OnSourceUnloadedListener;
import com.comscore.streaming.AdType;
import com.comscore.streaming.ContentType;
import com.comscore.streaming.ReducedRequirementsStreamingAnalytics;

import java.util.HashMap;
import java.util.Map;

public class ComScoreBitmovinAdapter {
    private static final String TAG = "ComScoreBitmovinAdapter";
    private static final String ASSET_DURATION_KEY = "ns_st_cl";

    private BitmovinPlayer bitmovinPlayer;
    private Map<String, String> metadata;
    private int contentType;
    private ReducedRequirementsStreamingAnalytics streamingAnalytics;
    private ComScoreState comScoreState = ComScoreState.STOPPED;
    private double currentAdDuration;
    private double currentAdOffset;

    public ComScoreBitmovinAdapter(BitmovinPlayer bitmovinPlayer, ComScoreMetadata comScoreMetadata) {
        this.bitmovinPlayer = bitmovinPlayer;
        metadata = comScoreMetadata.toDictionary();
        contentType = comScoreContentType(comScoreMetadata.getMediaType());
        this.streamingAnalytics = new ReducedRequirementsStreamingAnalytics();

        this.bitmovinPlayer.addEventListener(onPlaybackFinishedListener);
        this.bitmovinPlayer.addEventListener(onPlayingListener);
        this.bitmovinPlayer.addEventListener(onPausedListener);
        this.bitmovinPlayer.addEventListener(onErrorListener);
        this.bitmovinPlayer.addEventListener(onSourceUnloadedListener);
        this.bitmovinPlayer.addEventListener(onAdStartedListener);
        this.bitmovinPlayer.addEventListener(onAdFinishedListener);
    }

    public void updateMetadata(ComScoreMetadata metadata) {
        this.metadata = metadata.toDictionary();
        contentType = comScoreContentType(metadata.getMediaType());
    }

    private OnPlaybackFinishedListener onPlaybackFinishedListener = new OnPlaybackFinishedListener() {
        @Override
        public void onPlaybackFinished(PlaybackFinishedEvent playbackFinishedEvent) {
            stop();
        }
    };

    private OnPlayingListener onPlayingListener = new OnPlayingListener() {
        @Override
        public void onPlaying(PlayingEvent playingEvent) {
            if (bitmovinPlayer.isAd()) {
                transitionToAd(currentAdDuration, currentAdOffset);
            } else {
                transitionToVideoPlay();
            }
        }
    };

    private OnPausedListener onPausedListener = new OnPausedListener() {
        @Override
        public void onPaused(PausedEvent pausedEvent) {
            if (!bitmovinPlayer.isAd()){
                stop();
            }
        }
    };

    private OnErrorListener onErrorListener = new OnErrorListener() {
        @Override
        public void onError(ErrorEvent errorEvent) {
            stop();
        }
    };

    private OnSourceUnloadedListener onSourceUnloadedListener = new OnSourceUnloadedListener() {
        @Override
        public void onSourceUnloaded(SourceUnloadedEvent sourceUnloadedEvent) {
            stop();
        }
    };

    private OnAdStartedListener onAdStartedListener = new OnAdStartedListener() {
        @Override
        public void onAdStarted(AdStartedEvent adStartedEvent) {
            currentAdDuration = adStartedEvent.getDuration();
            currentAdOffset = adStartedEvent.getTimeOffset();
            transitionToAd(currentAdDuration, currentAdOffset);
        }
    };

    private OnAdFinishedListener onAdFinishedListener = new OnAdFinishedListener() {
        @Override
        public void onAdFinished(AdFinishedEvent adFinishedEvent) {
            transitionToVideoPlay();
        }
    };

    private int comScoreContentType(ComScoreMediaType mediaType) {
        switch (mediaType) {
            case LIVE:
                return ContentType.LIVE;
            case USER_GENERATED_SHORT_FORM_ON_DEMAND:
                return ContentType.USER_GENERATED_SHORT_FORM_ON_DEMAND;
            case USER_GENERATED_LONG_FORM_ON_DEMAND:
                return ContentType.USER_GENERATED_LONG_FORM_ON_DEMAND;
            case USER_GENERATED_LIVE:
                return ContentType.USER_GENERATED_LIVE;
            case BUMPER:
                return ContentType.BUMPER;
            case LONG_FORM_ON_DEMAND:
                return ContentType.LONG_FORM_ON_DEMAND;
            case SHORT_FORM_ON_DEMAND:
                return ContentType.SHORT_FORM_ON_DEMAND;
            default:
                return ContentType.OTHER;
        }
    }

    private synchronized void stop() {
        if (comScoreState != ComScoreState.STOPPED) {
            Log.d(TAG, "Stopping ComScore Tracking");
            comScoreState = ComScoreState.STOPPED;
            streamingAnalytics.stop();
        }
    }

    private synchronized void transitionToVideoPlay() {
        double duration = bitmovinPlayer.getDuration();
        if (Double.isInfinite(duration)) {
            duration = 0;
        } else {
            duration = duration * 1000;
        }
        metadata.put(ASSET_DURATION_KEY, String.valueOf((int) duration));

        if (comScoreState != ComScoreState.VIDEO) {
            stop();
            comScoreState = ComScoreState.VIDEO;
            Log.d(TAG, "ComScore Tracking Video Content");
            streamingAnalytics.playVideoContentPart(metadata, contentType);
        }
    }

    private synchronized void transitionToAd(double duration, double offset) {
        if (comScoreState != ComScoreState.ADVERTISEMENT) {
            stop();
            comScoreState = ComScoreState.ADVERTISEMENT;

            int adType;

            if (bitmovinPlayer.isLive()) {
                adType = AdType.LINEAR_LIVE;
            } else {
                if (offset == 0) {
                    adType = AdType.LINEAR_ON_DEMAND_PRE_ROLL;
                } else if (offset + duration == bitmovinPlayer.getDuration()) {
                    adType = AdType.LINEAR_ON_DEMAND_POST_ROLL;
                } else {
                    adType = AdType.LINEAR_ON_DEMAND_MID_ROLL;
                }
            }

            duration = duration * 1000;

            Map<String, String> adMap = new HashMap<>();
            adMap.put(ASSET_DURATION_KEY, String.valueOf((int) duration));
            Log.d(TAG, "ComScore Tracking Ad Play");
            streamingAnalytics.playVideoAdvertisement(adMap, adType);
        }
    }

    private enum ComScoreState {
        ADVERTISEMENT,
        STOPPED,
        VIDEO
    }
}

File: comscore/src/main/java/com/bitmovin/player/integration/comscore/ComScoreBitmovinAdapter.kt
package com.bitmovin.player.integration.comscore

import com.bitmovin.player.BitmovinPlayer
import com.bitmovin.player.api.event.listener.*
import com.bitmovin.player.integration.comscore.util.contentType
import com.bitmovin.player.integration.comscore.util.notifyHiddenEvent
import com.bitmovin.player.integration.comscore.util.notifyHiddenEvents
import com.comscore.streaming.AdvertisementMetadata
import com.comscore.streaming.AdvertisementType
import com.comscore.streaming.ContentMetadata
import com.comscore.streaming.StreamingAnalytics
import kotlin.properties.Delegates

class ComScoreBitmovinAdapter(private val bitmovinPlayer: BitmovinPlayer, private val configuration: ComScoreConfiguration, comScoreMetadata: ComScoreMetadata) {
    companion object {
        private const val ASSET_DURATION_KEY = "ns_st_cl"
    }

    private enum class ComScoreState {
        ADVERTISEMENT,
        STOPPED,
        VIDEO
    }

    private var metadataMap = comScoreMetadata.toMap().toMutableMap()
    private val streamingAnalytics = StreamingAnalytics()
    private var comScoreState = ComScoreState.STOPPED
    private var currentAdDuration = 0.0
    private var currentAdOffset = 0.0

    var suppressAdAnalytics = false

    var metadata: ComScoreMetadata by Delegates.observable(comScoreMetadata) { _, _, newMetadata ->
        metadataMap.clear()
        metadataMap.putAll(newMetadata.toMap())
    }

    init {
        BitLog.isEnabled = configuration.debug
        BitLog.d("Version ${BuildConfig.VERSION_NAME}")
        addEventListeners()
    }

    private fun addEventListeners() = with(bitmovinPlayer) {
        addEventListener(OnPlaybackFinishedListener {
            stop()
        })

        addEventListener(OnPlayingListener {
            if (bitmovinPlayer.isAd) {
                transitionToAd(currentAdDuration, currentAdOffset)
            } else {
                transitionToVideoPlay()
            }
        })

        addEventListener(OnPausedListener {
            if (!bitmovinPlayer.isAd) {
                stop()
            }
        })

        addEventListener(OnPausedListener {
            if (!bitmovinPlayer.isAd) {
                stop()
            }
        })

        addEventListener(OnErrorListener {
            stop()
        })

        addEventListener(OnSourceUnloadedListener {
            stop()
        })

        addEventListener(OnAdStartedListener {
            currentAdDuration = it.duration
            currentAdOffset = it.timeOffset
            transitionToAd(currentAdDuration, currentAdOffset)
        })

        addEventListener(OnAdFinishedListener {
            transitionToVideoPlay()
        })
    }

    @Synchronized
    private fun stop() {
        if (comScoreState != ComScoreState.STOPPED) {
            BitLog.d("Stopping ComScore tracking")
            comScoreState = ComScoreState.STOPPED
            streamingAnalytics.notifyPause()
        }
    }

    @Synchronized
    private fun transitionToVideoPlay() {
        var duration = bitmovinPlayer.duration
        if (duration.isInfinite()) {
            duration = 0.0
        } else {
            duration *= 1000
        }
        metadataMap[ASSET_DURATION_KEY] = duration.toInt().toString()
        if (comScoreState != ComScoreState.VIDEO) {
            stop()
            comScoreState = ComScoreState.VIDEO

            val contentMetadata = ContentMetadata.Builder().apply {
                mediaType(metadata.mediaType.contentType())
                customLabels(metadataMap)
            }.build()

            with(streamingAnalytics) {
                setMetadata(contentMetadata)
                notifyPlay()
            }

            BitLog.d("Starting ComScore video content tracking")
        }
    }

    @Synchronized
    private fun transitionToAd(duration: Double, offset: Double) {
        if (comScoreState != ComScoreState.ADVERTISEMENT) {
            stop()

            if (suppressAdAnalytics) {
                BitLog.d("Not tracking ad content as ad analytics is suppressed")
                return
            }

            comScoreState = ComScoreState.ADVERTISEMENT
            val adType = when {
                bitmovinPlayer.isLive -> AdvertisementType.LIVE
                offset == 0.0 -> AdvertisementType.ON_DEMAND_PRE_ROLL
                offset + duration == bitmovinPlayer.duration -> AdvertisementType.ON_DEMAND_POST_ROLL
                else -> AdvertisementType.ON_DEMAND_MID_ROLL
            }
            val durationValue = (duration * 1000).toInt().toString()
            val adMap = mapOf(ASSET_DURATION_KEY to durationValue)

            val adMetadata = AdvertisementMetadata.Builder().apply {
                mediaType(adType)
                customLabels(adMap)
            }.build()

            with(streamingAnalytics) {
                setMetadata(adMetadata)
                notifyPlay()
            }

            BitLog.d("Starting ComScore ad play tracking")
        }
    }

    fun setPersistentLabel(label: String, value: String) {
        notifyHiddenEvent(configuration.publisherId, label, value)
        BitLog.d("ComScore persistent label set: [$label:$value]")
    }

    fun setPersistentLabels(labels: Map<String, String>)  {
        notifyHiddenEvents(configuration.publisherId, labels)
        BitLog.d("ComScore persistent labels set: ${labels.map { "${it.key}:${it.value}" }}")
    }
}
